tensornetwork/network_operations.py

```diff
@@ -0,0 +1,575 @@
+      truncated_singular_values:
+        The vector of truncated singular values.
+  """
+  if not hasattr(node, 'backend'):
+    raise TypeError('Node {} of type {} has no `backend`'.format(
+        node, type(node)))
+
+  if node.axis_names and left_edge_name and right_edge_name:
+    left_axis_names = []
+    right_axis_names = [right_edge_name]
+    for edge in left_edges:
+      left_axis_names.append(node.axis_names[edge.axis1] if edge.node1 is node
+                             else node.axis_names[edge.axis2])
+    for edge in right_edges:
+      right_axis_names.append(node.axis_names[edge.axis1] if edge.node1 is node
+                              else node.axis_names[edge.axis2])
+    left_axis_names.append(left_edge_name)
+    center_axis_names = [left_edge_name, right_edge_name]
+  else:
+    left_axis_names = None
+    center_axis_names = None
+    right_axis_names = None
+
+  backend = node.backend
+
+  node.reorder_edges(left_edges + right_edges)
+  u, s, vh, trun_vals = backend.svd_decomposition(
+      node.tensor, len(left_edges), max_singular_values, max_truncation_err)
+  left_node = Node(
+      u, name=left_name, axis_names=left_axis_names, backend=backend.name)
+  singular_values_node = Node(
+      backend.diag(s),
+      name=middle_name,
+      axis_names=center_axis_names,
+      backend=backend.name)
+
+  right_node = Node(
+      vh, name=right_name, axis_names=right_axis_names, backend=backend.name)
+
+  for i, edge in enumerate(left_edges):
+    left_node.add_edge(edge, i)
+    edge.update_axis(i, node, i, left_node)
+  for i, edge in enumerate(right_edges):
+    # i + 1 to account for the new edge.
+    right_node.add_edge(edge, i + 1)
+    edge.update_axis(i + len(left_edges), node, i + 1, right_node)
+  connect(
+      left_node.edges[-1], singular_values_node.edges[0], name=left_edge_name)
+  connect(
+      singular_values_node.edges[1], right_node.edges[0], name=right_edge_name)
+  return left_node, singular_values_node, right_node, trun_vals
+
+
+def _reachable(node: BaseNode) -> Set[BaseNode]:
+  # Fastest way to get a single item from a set.
+  node_que = collections.deque()
+  seen_nodes = {node}
+  node_que.append(node)
+  while node_que:
+    node = node_que.popleft()
+    for e in node.edges:
+      for n in e.get_nodes():
+        if n is not None and n not in seen_nodes:
+          node_que.append(n)
+          seen_nodes.add(n)
+  return seen_nodes
+
+
+def reachable(nodes: Union[BaseNode, Union[List[BaseNode], Set[BaseNode]]]
+             ) -> Set[BaseNode]:
+  """
+  Computes all nodes reachable from `node` by connected edges.
+  Args:
+    node: A `BaseNode`
+  Returns:
+    A list of `BaseNode` objects that can be reached from `node`
+    via connected edges.
+  Raises:
+    ValueError: If an unknown value for `strategy` is passed.
+  """
+
+  if isinstance(nodes, BaseNode):
+    nodes = {nodes}
+
+  reachable_nodes = set()
+  for node in nodes:
+    reachable_nodes |= _reachable(node)
```

Comment:
This will run on O(n^2) instead of O(n) for the worst case if there is more than 1 node given to `reachable`

We'll fix it later, I'm just leaving this comment to show that I noticed it.